### DISCUSSION TITLE
Concurrent broadcasting

### DISCUSSION THREAD

**Discussion by aksu-104128** [miguelgrinberg/python-socketio#1291] [Ideas]
**Describe the solution you'd like**
The current broadcasting mechanism is to synchronously push messages to clients one after another in a for loop. If the number of clients is very large, this will be very slow. Can we use methods like GEVENT pool to broadcast concurrently?

**Comment by miguelgrinberg** [comment]
> The current broadcasting mechanism is to synchronously push messages to clients

I think the correct word here is "sequentially". I don't think broadcasting happens synchronously as you say. The only part that is sequential is the queuing of the events, because the actual writing to the client WebSockets happens in background tasks.
 
> Can we use methods like GEVENT pool to broadcast concurrently?

I'm not sure I understand What you are saying here. Gevent is single-threaded.

**Comment by aksu-104128** [comment]
@miguelgrinberg 

sorry for my bad english. For example, if we want to broadcast an event to 100 clients, can we use 100 greenlets, or use a thread pool to concurrent push to all clients instead of using one background thread to push sequentially？

**Reply by miguelgrinberg** [reply]
The `emit()` method does not do anything more than put what you want to send in a queue. It's fairly quick. The core logic for an emit is actually a [single line of code](https://github.com/miguelgrinberg/python-engineio/blob/main/src/engineio/socket.py#L75).

Each connected client has its own writer task that awakes when something shows up in the queue. This happens [here](https://github.com/miguelgrinberg/python-engineio/blob/main/src/engineio/socket.py#L194).

**Reply by aksu-104128** [reply]
Oh, I understand now, so basically each client has its own coroutine to read and write data. Thank you
